game-core/src/test/java/games/strategy/triplea/delegate/power/calculator/TotalPowerAndTotalRollsTest.java

```diff
@@ -92,62 +782,2542 @@ void testDefending() {
+      final GameData gameData = givenGameData().build();
+      assertThat(
+          TotalPowerAndTotalRolls.getTotalAaAttacks(
+              Map.of(
+                  givenUnit("test", BattleType.NORMAL, gameData, OFFENSE, 1, -1),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(-1).build(),
+                  givenUnit("test2", BattleType.NORMAL, gameData, OFFENSE, 1, 10),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(10).build()),
+              List.of(mock(Unit.class), mock(Unit.class), mock(Unit.class))),
+          is(3));
+    }
+
+    @Test
+    void rollsOfNonInfiniteUnitEqualsAttack() throws MutableProperty.InvalidValueException {
+      final GameData gameData = givenGameData().build();
+      assertThat(
+          TotalPowerAndTotalRolls.getTotalAaAttacks(
+              Map.of(
+                  givenUnit("test", BattleType.NORMAL, gameData, OFFENSE, 1, 1),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(1).build()),
+              List.of(mock(Unit.class), mock(Unit.class), mock(Unit.class))),
+          is(1));
+    }
+
+    @Test
+    void rollsOfNonInfiniteUnitGreaterThanTargetCountMeansAttackEqualsTarget()
+        throws MutableProperty.InvalidValueException {
+      final GameData gameData = givenGameData().build();
+      assertThat(
+          TotalPowerAndTotalRolls.getTotalAaAttacks(
+              Map.of(
+                  givenUnit("test", BattleType.NORMAL, gameData, OFFENSE, 1, 2),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(2).build(),
+                  givenUnit("test2", BattleType.NORMAL, gameData, OFFENSE, 1, 2),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(2).build()),
+              List.of(mock(Unit.class), mock(Unit.class), mock(Unit.class))),
+          is(3));
+    }
+
+    @Test
+    void overstackUnitCanCauseAttackToGoOverTargetCount()
+        throws MutableProperty.InvalidValueException {
+      final GameData gameData = givenGameData().build();
+      final Unit overstackUnit = givenUnit("test", BattleType.NORMAL, gameData, OFFENSE, 1, 3);
+      overstackUnit.getUnitAttachment().getPropertyOrThrow(MAY_OVERSTACK_AA).setValue(true);
+
+      assertThat(
+          TotalPowerAndTotalRolls.getTotalAaAttacks(
+              Map.of(
+                  overstackUnit,
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(2).build(),
+                  givenUnit("test2", BattleType.NORMAL, gameData, OFFENSE, 1, 3),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(3).build(),
+                  givenUnit("test3", BattleType.NORMAL, gameData, OFFENSE, 1, -1),
+                  TotalPowerAndTotalRolls.builder().totalPower(1).totalRolls(-1).build()),
+              List.of(mock(Unit.class), mock(Unit.class), mock(Unit.class))),
+          is(5));
```

Comment:
Yeah, it is dealing with rolls only.  I added some description.  I thought the name of the method would be ok but I think that the terminology is wrong.  There is "rolls", "power", "strength", "attacks", etc.  I think "rolls" always means "rolls" and the others all mean the same, except when "attacks" means "rolls" (such as "maxAAattacks" and this method "getTotalAaAttacks").  What is the current term: "power", "strength", "attacks", or something else?

Reply:
IMO always err on making tests stupid simple to understand and spell out explicitly what is being tested. Humbly the test name here did not inform me it is about rolls. Also like javadocs and any other commentary, remember it's not the ritual that is helpful, so the commentary/description can stay very terse, does not have to be full sentences - the audience are future developers that either are hunting down a bug and looking for a test gap, or just fixed a bug and are fixing broken tests, or changed the code and now are looking to fix a number of broken tests.

Thank you for adding the description. I'll take a look in more detail I hope soon.

Re: terms

power -> typically this is strength times number of rolls. 
strength -> the value on the die to hit
attacks -> inconsistent name for 'number of rolls'

So, a bomber has power and strength 4 with 1 roll, with heavy bomber tech the power goes up to 8, strength stays the same. With LHTR heavy bomber, it's less defined, power is mostly used by AI to approximate unit capability, so the LHTR power would probably be 5.28   ((1 - (2/6 * 2/6))*6)   (Inverse of the odds of not hitting times number of sides on dice)

Reply:
> power -> typically this is strength times number of rolls.
> strength -> the value on the die to hit
> attacks -> inconsistent name for 'number of rolls'

Then these files are all misnamed.  `TotalPowerAndTotalRolls` should be renamed to `TotalStrengthAndTotalRolls` since the "power" that it is calculated is the value on the die to hit.  `getTotalAaAttacks` should be `getTotalAaRolls`, `getMaxAaAttackAndDiceSides` should be `getMaxAaStrengthAndDiceSides`, etc.

Reply:
Hmm, good point. My understanding might have been inconsistent, or it's perhaps just not consistent. I'd recommend doing a survey of the terms and how they are typically used. This is something that would do well to be added to the [glossary](https://github.com/triplea-game/triplea/blob/master/docs/glossary.md)